File: design-patterns/java/singleton/DoubleCheckedLockingSingleton.java
class DoubleCheckedSingleton {
    // The single instance, initially null, marked as volatile
    private static volatile DoubleCheckedSingleton instance;
    
    // Private constructor to prevent instantiation
    private DoubleCheckedSingleton() {}
    
    // Public method to get the instance
    public static DoubleCheckedSingleton getInstance() {
        // First check (not synchronized)
        if (instance == null) {
            // Synchronize on the class object
            synchronized (DoubleCheckedSingleton.class) {
                // Second check (synchronized)
                if (instance == null) {
                    // Create the instance
                    instance = new DoubleCheckedSingleton();
                }
            }
        }
        // Return the instance (either newly created or existing)
        return instance;
    }
}

File: design-patterns/csharp/singleton/DoubleCheckedLockingSingleton.cs
namespace Singleton
{
    public class DoubleCheckedLockingSingleton
    {
        private static volatile DoubleCheckedLockingSingleton instance;
        private static readonly object lockObj = new object();
        private DoubleCheckedLockingSingleton() { }
        public static DoubleCheckedLockingSingleton GetInstance()
        {
            if (instance == null)
            {
                lock (lockObj)
                {
                    if (instance == null)
                    {
                        instance = new DoubleCheckedLockingSingleton();
                    }
                }
            }
            return instance;
        }
    }
} 